### PR TITLE
[PLGN-355] Confluence - Update auth and API Library

### PR DESCRIPTION
**Pull request by ablakley-r7**
### Description

Describe the proposed changes:

  - Update to use API Token as basic auth with password is deprecated
  - Update to use new API library as old library is no longer maintained
  - Update to make use of new library functionality (get_page_by_id, get_page_content_by_id)

## PR Requirements

Developers, verify you have completed the following items by checking them off:

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `insight-plugin validate` and make sure everything passes
2. Run the assessment tool: `insight-plugin run -A`. For single action validation: `insight-plugin run tests/{file}.json -A`
3. Copy (`insight-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

**Review by igorski-r7 — CHANGES_REQUESTED** [review]

File: plugins/confluence/komand_confluence/util/api.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        if page_exists:
+            logging.info("Updating page...")
+            page_id = self.get_page_id(title=title, space=space)
+            return self.confluence.update_page(page_id=page_id, title=title, body=content)
+        else:
+            logging.info("Creating a new page...")
+            return self.confluence.create_page(title=title, body=content, space=space)
+
+    @exception_handler
+    def page_exists(self, space: str, title: str):
```
Comment (igorski-r7) [inline plugins/confluence/komand_confluence/util/api.py:67]: Could you please add couple more type hints?

File: plugins/confluence/komand_confluence/util/util.py [hunk 2]
```diff
@@ -12,3 +21,60 @@ def normalize_page(p):
         p["current"] = True
 
     return p
+
+
+def extract_page_data(page):
+    url_base = page.get("_links", {}).get("base")
+    endpoint = page.get("_links", {}).get("webui")
+    home_page = page.get("space", {}).get("_expandable").get("homepage", "")
```
Comment (igorski-r7) [inline plugins/confluence/komand_confluence/util/util.py:29]:
```suggestion
    home_page = page.get("space", {}).get("_expandable", {}).get("homepage", "")
```

File: plugins/confluence/komand_confluence/util/util.py [hunk 3]
```diff
@@ -12,3 +21,60 @@ def normalize_page(p):
...
+        "version": f'{page.get("version", {}).get("number")}',
+        "homePage": is_home_page,
+        "id": page_id,
+        "current": page.get("status").lower() == "current",
+        "contentStatus": page.get("status"),
+        "modified": page.get("version", {}).get("when"),
+    }
+
+
+def exception_handler(func):
```
Comment (igorski-r7) [inline plugins/confluence/komand_confluence/util/util.py:53]:
```suggestion
def exception_handler(func: Callable):
```

**Review by igorski-r7 — APPROVED** [review]

File: plugins/confluence/komand_confluence/actions/get_page/action.py [hunk 4]
```diff
@@ -1,46 +1,29 @@
...
-                "id": "19726355",
-                "current": "true",
-                "contentStatus": "current",
-                "modified": "20161024T20:19:23Z",
-            },
-        }
+        title = params.get("page")
+        space = params.get("space")
+        page_id = self.connection.client.get_page_id(title=title, space=space)
+        if page_id:
```
Comment (dlaverty-r7) [inline plugins/confluence/komand_confluence/actions/get_page/action.py:23]: Is it worth logging the page id here for troubleshooting?

**Review by dlaverty-r7 — APPROVED** [review]

**Comment by dlaverty-r7** [comment]
@ablakley-r7 , can you double check the pluginvalidator issue prior to merge please?

**Comment by ablakley-r7** [comment]
Validator checking connection details, unrecognised major bump. Will investigate validator.

**Pull request merged**